lib/std/map.zig

```diff
@@ -0,0 +1,872 @@
+            return self.node_count;
+        }
+
+        pub fn insert(self: *Self, key: K, value: V) !void {
+            if (self.root == null) {
+                self.root = try Node.init(self.allocator, key, value);
+                self.root.?.color = Color.Black;
+                self.node_count = 1;
+                return;
+            }
+
+            var current = self.root.?;
+            var parent: ?*Node = null;
+
+            while (true) {
+                parent = current;
+                if (self.compare(key, current.key) < 0) {
+                    if (current.left) |left| {
+                        current = left;
+                    } else {
+                        current.left = try Node.init(self.allocator, key, value);
+                        current.left.?.parent = current;
+                        self.insertFixup(current.left.?);
+                        self.node_count += 1;
+                        return;
+                    }
+                } else if (self.compare(key, current.key) > 0) {
+                    if (current.right) |right| {
+                        current = right;
+                    } else {
+                        current.right = try Node.init(self.allocator, key, value);
+                        current.right.?.parent = current;
+                        self.insertFixup(current.right.?);
+                        self.node_count += 1;
+                        return;
+                    }
+                } else {
+                    current.value = value;
+                    return;
+                }
+            }
+        }
+
+        pub fn getOrPut(self: *Self, key: K, default: V) !*V {
+            if (self.findNode(key)) |node| {
+                return &node.value;
+            } else {
+                try self.insert(key, default);
+                return &self.findNode(key).?.value;
+            }
+        }
+
+        pub fn get(self: *const Self, key: K) ?V {
+            if (self.findNode(key)) |node| {
+                return node.value;
+            }
+            return null;
+        }
+
+        pub fn getPtr(self: *Self, key: K) ?*V {
+            if (self.findNode(key)) |node| {
+                return &node.value;
+            }
+            return null;
+        }
+
+        pub fn contains(self: *const Self, key: K) bool {
+            return self.findNode(key) != null;
+        }
+
+        pub fn count(self: *const Self, key: K) usize {
+            return if (self.contains(key)) 1 else 0;
+        }
+
+        pub fn erase(self: *Self, key: K) bool {
+            if (self.findNode(key)) |node| {
+                self.deleteNode(node);
+                self.node_count -= 1;
+                return true;
+            }
+            return false;
+        }
+
+        pub fn iterator(self: *const Self) Iterator {
+            return Iterator{ .current = self.minimum(self.root) };
+        }
+
+        fn compare(self: *const Self, a: K, b: K) i32 {
```

Comment:
This compare function should probably be provided as an optional argument to the map, to allow custom comparisons. Similar to how the hash maps do it.